------------------------------------------------------------------------
r1672295 | rhuijben | 2015-04-09 07:31:12 -0400 (Thu, 09 Apr 2015) | 15 lines

Optimize STMT_SELECT_EXTERNALS_DEFINED when using Sqlite 3.8.9, by
adding two more rows to the sqlite_stat1 table.

This fixes a test failure in wc-queries-test.c, but actual users most
likely don't notice a difference as a tablescan on an EXTERNALS tables
index is not that expensive, given that most working copies don't have
many externals.

* subversion/libsvn_wc/wc-metadata.sql
  (STMT_INSTALL_SCHEMA_STATISTICS): Add two rows.

* subversion/tests/libsvn_wc/wc-queries-test.c
  (test_schema_statistics): Add a dummy externals row to allow
    verifying schema.


Index: git/subversion/tests/libsvn_wc/wc-queries-test.c
===================================================================
--- git.orig/subversion/tests/libsvn_wc/wc-queries-test.c	2015-08-16 00:14:10.791505879 +0200
+++ git/subversion/tests/libsvn_wc/wc-queries-test.c	2015-08-16 00:14:10.787505827 +0200
@@ -833,6 +833,15 @@
                    "VALUES (1, 'subdir', '', 1, 'normal', 'dir', '', '')",
                    NULL, NULL, NULL));
 
+  SQLITE_ERR(
+      sqlite3_exec(sdb,
+                   "INSERT INTO EXTERNALS (wc_id, local_relpath,"
+                   "                       parent_relpath, repos_id,"
+                   "                       presence, kind, def_local_relpath,"
+                   "                       def_repos_relpath) "
+                   "VALUES (1, 'subdir', '', 1, 'normal', 'dir', '', '')",
+                   NULL, NULL, NULL));
+
   /* These are currently not necessary for query optimization, but it's better
      to tell Sqlite how we intend to use this table anyway */
   SQLITE_ERR(
Index: git/subversion/libsvn_wc/wc-metadata.sql
===================================================================
--- git.orig/subversion/libsvn_wc/wc-metadata.sql	2015-08-16 00:14:10.791505879 +0200
+++ git/subversion/libsvn_wc/wc-metadata.sql	2015-08-16 00:14:10.787505827 +0200
@@ -624,6 +624,11 @@
 INSERT INTO sqlite_stat1(tbl, idx, stat) VALUES
     ('EXTERNALS','I_EXTERNALS_DEFINED',                 '100 100 3 1');
 
+INSERT OR REPLACE INTO sqlite_stat1(tbl, idx, stat) VALUES
+    ('EXTERNALS','sqlite_autoindex_EXTERNALS_1',        '100 100 1');
+INSERT OR REPLACE INTO sqlite_stat1(tbl, idx, stat) VALUES
+    ('EXTERNALS','I_EXTERNALS_DEFINED',                 '100 100 3 1');
+
 /* sqlite_autoindex_WORK_QUEUE_1 doesn't exist because WORK_QUEUE is
    a INTEGER PRIMARY KEY AUTOINCREMENT table */
 
